File: FlashQuizTests/NetworkSessionManagerTests.m
//
//  NetworkSessionManagerTests.m
//  SmallSharpTools
//
//  Created by Brennan Stehling on 7/10/16.
//  Copyright © 2016 FlashQuiz. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "AppConfiguration.h"
#import "NetworkSessionManager.h"
#import "Question.h"

@interface NetworkSessionManager (NetworkSessionManagerTests)

- (nonnull NSURL *)appendQueryParameters:(nullable NSDictionary *)params toURL:(nonnull NSURL *)url;

@end

@interface NetworkSessionManagerTests : XCTestCase

@end

@implementation NetworkSessionManagerTests

- (void)setUp {
    [super setUp];
}

- (void)tearDown {
    [super tearDown];
}

- (void)testFetchingJSON {
    NSURL *jsonURL = [NSURL URLWithString:@"https://sst-robots.s3.amazonaws.com/data.json"];
    NetworkSessionManager *manager = [NetworkSessionManager new];

    XCTestExpectation *expectation = [self expectationWithDescription:@"Fetching JSON"];

    [manager fetchJSONWithURL:jsonURL params:nil withCompletionBlock:^(id  _Nullable json, NSError * _Nullable error) {
        XCTAssertTrue([NSThread isMainThread], @"Must be main thread");
        XCTAssertNil(error, @"Error is not expected");
        XCTAssertNotNil(json, @"JSON is expected");
        XCTAssertTrue([json isKindOfClass:[NSDictionary class]], @"Must be a dictionary");
        [expectation fulfill];
    }];

    NSTimeInterval timeout = 10; // 10 seconds
    [self waitForExpectationsWithTimeout:timeout handler:^(NSError * _Nullable error) {
        // do nothing
    }];
}

- (void)testFetchingImage {
    NSURL *imageURL = [NSURL URLWithString:@"https://sst-robots.s3.amazonaws.com/robot.jpg"];
    NetworkSessionManager *manager = [NetworkSessionManager new];

    XCTestExpectation *expectation = [self expectationWithDescription:@"Fetching Image"];

    [manager fetchImageWithURL:imageURL params:nil withCompletionBlock:^(UIImage * _Nullable image, NSError * _Nullable error) {
        XCTAssertTrue([NSThread isMainThread], @"Must be main thread");
        XCTAssertNil(error, @"Error is not expected");
        XCTAssertNotNil(image, @"Image is expected");
        [expectation fulfill];
    }];

    NSTimeInterval timeout = 10; // 10 seconds
    [self waitForExpectationsWithTimeout:timeout handler:^(NSError * _Nullable error) {
        // do nothing
    }];
}

- (void)testFetchingImages {
    NSURL *fileURL = [AppConfiguration questionsFileURL];
    NSArray<Question *> *questions = [Question questionsWithFileURL:fileURL];
    NetworkSessionManager *manager = [NetworkSessionManager new];

    XCTestExpectation *expectation = [self expectationWithDescription:@"Fetching Images"];

    __block NSUInteger count = 0;

    Question *question = questions.firstObject;
    XCTAssertNotNil(question);
    XCTAssert(question.answers.count > 0);

    for (NSString *answer in question.answers) {
        NSURL *imageURL = [NSURL URLWithString:answer];
        [manager fetchImageWithURL:imageURL params:nil withCompletionBlock:^(UIImage * _Nullable image, NSError * _Nullable error) {
            XCTAssertNil(error, @"Error is not expected");
            XCTAssertNotNil(image, @"Image is expected");
            count++;

            if (count == question.answers.count) {
                [expectation fulfill];
            }
        }];
    }

    NSTimeInterval timeout = 10; // 10 seconds
    [self waitForExpectationsWithTimeout:timeout handler:^(NSError * _Nullable error) {
        // do nothing
    }];
}

- (void)testNoCompletionBlocks {
    NSURL *anyURL = nil;
    NetworkSessionManager *manager = [NetworkSessionManager new];

    NSURLSessionTask *task1 = [manager fetchJSONWithURL:anyURL params:nil withCompletionBlock:nil];
    NSURLSessionTask *task2 = [manager fetchImageWithURL:anyURL params:nil withCompletionBlock:nil];
    NSURLSessionTask *task3 = [manager fetchDataWithURL:anyURL params:nil withCompletionBlock:nil];

    XCTAssertNil(task1);
    XCTAssertNil(task2);
    XCTAssertNil(task3);
}

- (void)testAppendingParameters1 {
    NSURL *servicesURL = [NSURL URLWithString:@"https://api.FlashQuiz.com/search"];
    NSDictionary *params = @{ @"text" : @"carrot", @"format" : @"json" };
    NetworkSessionManager *manager = [NetworkSessionManager new];

    NSURL *appendedURL = [manager appendQueryParameters:params toURL:servicesURL];
    XCTAssertNotNil(appendedURL);
    XCTAssertTrue([appendedURL.absoluteString rangeOfString:servicesURL.absoluteString].location != NSNotFound);
    for (NSString *key in params.allKeys) {
        NSString *pair = [NSString stringWithFormat:@"%@=%@", key, params[key]];
        XCTAssertTrue([appendedURL.query rangeOfString:pair].location != NSNotFound);
    }
}

- (void)testAppendingParameters2 {
    NSURL *servicesURL = [NSURL URLWithString:@"https://api.FlashQuiz.com/search"];
    NSDictionary *params = @{ @"text" : @"carrot", @"format" : @"json", @"pageSize" : @(20) };
    NetworkSessionManager *manager = [NetworkSessionManager new];

    NSURL *appendedURL = [manager appendQueryParameters:params toURL:servicesURL];
    XCTAssertNotNil(appendedURL);
    XCTAssertTrue([appendedURL.absoluteString rangeOfString:servicesURL.absoluteString].location != NSNotFound);
    for (NSString *key in params.allKeys) {
        NSString *pair = [NSString stringWithFormat:@"%@=%@", key, params[key]];
        XCTAssertTrue([appendedURL.query rangeOfString:pair].location != NSNotFound);
    }
}

@end

File: FlashQuizSwiftTests/NetworkSessionManagerTests.swift
//
//  NetworkSessionManagerTests.swift
//  FlashQuiz
//
//  Created by Brennan Stehling on 7/17/16.
//  Copyright © 2016 SmallSharpTools. All rights reserved.
//

import XCTest

class NetworkSessionManagerTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
    }
    
    override func tearDown() {
        super.tearDown()
    }
    
    func testFetchingJSON() {
        let expectation = self.expectation(description: "Fetching JSON")

        guard let jsonURL = URL(string: "https://sst-robots.s3.amazonaws.com/data.json") else {
            XCTFail("Unable to create URL")
            return
        }

        let manager = NetworkSessionManager()
        manager.fetchJSON(with: jsonURL, params: nil) { (json, error) in
            XCTAssertTrue(Thread.isMainThread, "Must be main thread")
            XCTAssertNil(error, "Error is not expected")
            XCTAssertNotNil(json, "JSON is expected")
            if let json = json {
                XCTAssertTrue((json as AnyObject).isKind(of: NSDictionary.self), "Must be a dictionary")
            }

            expectation.fulfill()
        }

        let timeout: TimeInterval = 10
        self.waitForExpectations(timeout: timeout) { (error) in
            // do nothing
        }
    }

    func testFetchingImage() {
        let expectation = self.expectation(description: "Fetching Image")

        guard let imageURL = URL(string: "https://sst-robots.s3.amazonaws.com/robot.jpg") else {
            XCTFail("Unable to create URL")
            return
        }

        let manager = NetworkSessionManager()
        manager.fetchImage(with: imageURL, params: nil) { (image, error) in
            XCTAssertTrue(Thread.isMainThread, "Must be main thread")
            XCTAssertNil(error, "Error is not expected")
            XCTAssertNotNil(image, "Image is expected")

            expectation.fulfill()
        }

        let timeout: TimeInterval = 10
        self.waitForExpectations(timeout: timeout) { (error) in
            // do nothing
        }
    }

    func testFetchingImages() {
        let fileURL = AppConfiguration.questionsFileURL()
        let questions = Question.questions(withFileURL: fileURL)
        let manager = NetworkSessionManager()

        let expectation = self.expectation(description: "Fetching Images")

        XCTAssertNotNil(questions)
        if let questions = questions {
            let question = questions.first
            XCTAssertNotNil(question)

            var count: Int = 0

            if let question = question {
                XCTAssert(question.answers.count > 0)
                for answer in question.answers {
                    print("Answer: \(answer)")
                    let imageURL = URL(string: answer)
                    if let imageURL = imageURL {
                        manager.fetchImage(with: imageURL, params: nil, withCompletionBlock: { (image, error) in
                            XCTAssertNil(error, "Error is not expected")
                            XCTAssertNotNil(image, "Image is expected")
                            count += 1

                            if count == question.answers.count {
                                expectation.fulfill()
                            }
                        })

                    }
                }
            }
        }
        
        let timeout: TimeInterval = 10
        self.waitForExpectations(timeout: timeout) { (error) in
            // do nothing
        }
    }

    func testNoCompletionBlocks() {
        guard let anyURL = URL(string: "http://localhost") else {
            XCTFail()
            return
        }
        let manager = NetworkSessionManager()

        let task1 = manager.fetchJSON(with: anyURL, params: nil, withCompletionBlock: nil)
        let task2 = manager.fetchImage(with: anyURL, params: nil, withCompletionBlock: nil)
        let task3 = manager.fetchData(with: anyURL, params: nil, withCompletionBlock: nil)

        XCTAssertNil(task1)
        XCTAssertNil(task2)
        XCTAssertNil(task3)
    }

}
